code/exercises/affine-cipher/example.scm

```diff
@@ -0,0 +1,89 @@
+(import (rnrs))
+
+
+(define (encode key text)
+  ((get-translator key) (string-downcase text) 'encode))
+
+(define (decode key text)
+  ((get-translator key) text 'decode))
+
+;;; Private
+
+(define (atbash-format lst)
+  (let* ((chunks (windows lst 5))
+         (l (cons (car chunks)
+                  (map (lambda (chars) (cons #\space chars))
+                       (cdr chunks)))))
+    (list->string (fold-left append '() l))))
+
+(define (get-translator key)
+   (translator-factory
+    (cdr (assoc 'a key))
+    (cdr (assoc 'b key))))
+
+;; Work Horse - return a translator tuned to a specific affine key
+(define (translator-factory a b)
+  (when (not (coprime? a 26))
+    (error 'affine-cipher "a and m must be coprime" a))
+  (let* (;; generate a char pair from an alphabetic index
+         (index->pair
+          (lambda (idx)            ;; (a * x + b) % m
+            (let ((cipher-idx (modulo (+ (* a idx) b) 26)))
+              `(,(idx->char idx) . ,(idx->char cipher-idx)))))
+         ;; build an alist mapping the alphabet to the 'cipherbet'
+         (encode-map (map index->pair (iota 26)))
+         ;; build an alist mapping the 'cipherbet' to the alphabet
+         (decode-map (map riap encode-map))
+         ;; en/decrypt text and return a char list
+         (translate
+          (lambda (text tr-map)
+            (filter
+             ;; identity - to remove #f from the list
+             (lambda (x) x)
+             (map (lambda (c)
+                    (cond ((assoc c tr-map) => cdr)
+                          ((char-numeric? c) c)
+                          (else #f)))
+                  (string->list text))))))
+    (lambda (text msg)
+      (case msg
+        ((encode) (atbash-format (translate text encode-map)))
+        ((decode) (list->string (translate text decode-map)))
+        (else (error 'translator-factory "invalid message" msg))))))
+
+;;; Generic Helper Functions
+
+(define (coprime? x y)
+  (= 1 (gcd x y)))
+
+(define (idx->char idx)
+  (integer->char (+ idx (char->integer #\a))))
+
+(define (riap pair)
+  (when (not (pair? pair))
+    (error 'riap "wrong type arg is not a pair" pair))
+  (cons (cdr pair) (car pair)))
+
+;;; To Handle Formatting
+
+(define (windows lst size)
+  (let *windows* ((l lst) (acc '()))
+    (if (>= (length l) size)
+        (let ((window (take l size)))
+          (*windows* (drop l size) (cons window acc)))
+        (if (null? l)
+            (reverse acc)
+            (reverse (cons l acc))))))
```

Comment:
Basically stolen from the Rust standard libraries.  It is just such a good name

Reply:
Chunks in haskell?

Reply:
I think I have used that before.

Do you prefer the name chunks?

Reply:
chunks sounds kind of gross haha. windows has a nice ring to it, but I'm not sure I love either name to describe it... windows >> chunks for sure though